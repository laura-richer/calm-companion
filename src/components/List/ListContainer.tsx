import React from 'react';
import { View } from 'react-native';
import ListItem from './ListItem';
import styles from './styles';
import { ListContainerProps } from './types';
import { ListItemType } from '~types/list';

const ListContainer = ({
  listItems,
  handleItemClick,
  handleHintClick,
  accessabilityLabel,
}: ListContainerProps) => {
  return (
    <View style={styles.listContainer} accessibilityLabel={accessabilityLabel}>
      {listItems.map((listItem: ListItemType) => (
        <ListItem
          key={listItem.id}
          item={listItem}
          handleItemClick={handleItemClick}
          handleHintClick={handleHintClick}
        />
      ))}
    </View>
  );
};

export default ListContainer;
