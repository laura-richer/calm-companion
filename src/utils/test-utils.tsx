import React, { PropsWithChildren } from 'react';
import { NavigationContainer } from '@react-navigation/native';
import { render, RenderOptions } from '@testing-library/react-native';
import { Provider as StoreProvider } from 'react-redux';
import { EnhancedStore, configureStore } from '@reduxjs/toolkit';
import { SafeAreaProvider } from 'react-native-safe-area-context';
import { RootState } from '~store/index';
import navigationReducer, {
  initialState as navigationInitialState,
} from '~store/slices/navigation';
import quickHelpReducer, { initialState as quickHelpInitialState } from '~store/slices/quickHelp';

type DeepPartial<T> = {
  [K in keyof T]?: T[K] extends object ? DeepPartial<T[K]> : T[K];
};

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: Partial<RootState>;
  store?: EnhancedStore;
}

export const renderWithProviders = (
  ui: React.ReactElement,
  { preloadedState = {}, store = createTestStore(preloadedState) }: ExtendedRenderOptions = {}
) => {
  const wrapper = ({ children }: PropsWithChildren<{}>): JSX.Element => {
    return (
      <StoreProvider store={store}>
        <SafeAreaProvider>
          <NavigationContainer>{children}</NavigationContainer>
        </SafeAreaProvider>
      </StoreProvider>
    );
  };

  return {
    ...render(ui, {
      wrapper,
    }),
  };
};

const _createStore = (preloadedState: Partial<RootState> = {}): EnhancedStore => {
  return configureStore({
    reducer: {
      navigation: navigationReducer,
      quickHelp: quickHelpReducer,
    },
    preloadedState,
    middleware: getDefaultMiddleware =>
      getDefaultMiddleware({
        immutableCheck: { warnAfter: 256 },
        serializableCheck: { warnAfter: 256 },
      }),
  });
};

export const createTestStore = (preloadedState: DeepPartial<RootState> = {}): EnhancedStore => {
  const state: RootState = {
    // @ts-ignore
    navigation: {
      ...navigationInitialState,
      ...(preloadedState.navigation || {}),
    },
    // @ts-ignore
    quickHelp: {
      ...quickHelpInitialState,
      ...(preloadedState.quickHelp || {}),
    },
  };
  const store: EnhancedStore = _createStore(state);
  const origDispatch = store.dispatch;
  store.dispatch = jest.fn(origDispatch);
  return store;
};
