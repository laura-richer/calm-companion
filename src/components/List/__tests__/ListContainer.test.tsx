import React from 'react';
import { screen } from '@testing-library/react-native';
import { renderWithProviders } from '~utils/test-utils';
import ListContainer from '../ListContainer';
import { ListItemType } from '~types/list';

const listItems: ListItemType[] = [
  {
    id: 1,
    title: 'Item title',
    slug: 'item-title',
    done: false,
  },
  {
    id: 2,
    title: 'Item title 2',
    slug: 'item-title-2',
    done: false,
  },
  {
    id: 3,
    title: 'Item title 3',
    slug: 'item-title-3',
    done: false,
  },
];

const mockItemClick = jest.fn();
const mockHintClick = jest.fn();

const _renderDefault = () => {
  return renderWithProviders(
    <ListContainer
      listItems={listItems}
      handleItemClick={mockItemClick}
      handleHintClick={mockHintClick}
      accessabilityLabel="List accessability label"
    />
  );
};

describe('list container', () => {
  beforeEach(() => {
    _renderDefault();
  });

  test('matches snapshot', () => {
    expect(screen.toJSON()).toMatchSnapshot();
  });

  test('renders the correct amount of list items', () => {
    const renderedListItems = screen.getAllByAccessibilityHint('Tap to mark as done');
    expect(renderedListItems).toHaveLength(3);
  });
});
