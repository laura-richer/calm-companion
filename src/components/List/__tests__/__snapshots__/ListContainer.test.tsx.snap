// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`list container matches snapshot 1`] = `
<View
  accessibilityLabel="List accessability label"
  style={
    {
      "display": "flex",
      "gap": 20,
    }
  }
>
  <View
    style={
      [
        {
          "shadowColor": "#8b8490",
          "shadowOffset": {
            "height": 2,
            "width": 0,
          },
          "shadowOpacity": 0.25,
          "shadowRadius": 8,
        },
        {
          "backgroundColor": "#ffffff",
          "borderRadius": 8,
        },
      ]
    }
  >
    <BVLinearGradient
      angle={170}
      angleCenter={
        {
          "x": 0.75,
          "y": 0.75,
        }
      }
      colors={
        [
          4294967295,
          4294967295,
        ]
      }
      endPoint={
        {
          "x": 0.5,
          "y": 1,
        }
      }
      locations={null}
      startPoint={
        {
          "x": 0.5,
          "y": 0,
        }
      }
      style={
        {
          "alignItems": "center",
          "borderRadius": 8,
          "display": "flex",
          "flexDirection": "row",
          "gap": 10,
          "paddingHorizontal": 20,
          "paddingVertical": 10,
        }
      }
      useAngle={true}
    >
      <View
        accessibilityHint="Tap to mark as done"
        accessibilityLabel="Toggle Item title done state"
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        role="button"
        style={
          {
            "alignItems": "center",
            "display": "flex",
            "flex": 1,
            "flexDirection": "row",
            "gap": 10,
          }
        }
      >
        <View
          accessibilityLabel="Item title is not done"
          style={
            {
              "borderColor": "#3f7b82",
              "borderRadius": 50,
              "borderWidth": 2,
              "height": 25,
              "width": 25,
            }
          }
        />
        <Text
          style={
            [
              {
                "color": "#3f7b82",
              },
              {
                "fontFamily": "Poppins-Bold",
                "fontSize": 14,
              },
              null,
            ]
          }
        >
          Item title
        </Text>
      </View>
    </BVLinearGradient>
  </View>
  <View
    style={
      [
        {
          "shadowColor": "#8b8490",
          "shadowOffset": {
            "height": 2,
            "width": 0,
          },
          "shadowOpacity": 0.25,
          "shadowRadius": 8,
        },
        {
          "backgroundColor": "#ffffff",
          "borderRadius": 8,
        },
      ]
    }
  >
    <BVLinearGradient
      angle={170}
      angleCenter={
        {
          "x": 0.75,
          "y": 0.75,
        }
      }
      colors={
        [
          4294967295,
          4294967295,
        ]
      }
      endPoint={
        {
          "x": 0.5,
          "y": 1,
        }
      }
      locations={null}
      startPoint={
        {
          "x": 0.5,
          "y": 0,
        }
      }
      style={
        {
          "alignItems": "center",
          "borderRadius": 8,
          "display": "flex",
          "flexDirection": "row",
          "gap": 10,
          "paddingHorizontal": 20,
          "paddingVertical": 10,
        }
      }
      useAngle={true}
    >
      <View
        accessibilityHint="Tap to mark as done"
        accessibilityLabel="Toggle Item title 2 done state"
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        role="button"
        style={
          {
            "alignItems": "center",
            "display": "flex",
            "flex": 1,
            "flexDirection": "row",
            "gap": 10,
          }
        }
      >
        <View
          accessibilityLabel="Item title 2 is not done"
          style={
            {
              "borderColor": "#3f7b82",
              "borderRadius": 50,
              "borderWidth": 2,
              "height": 25,
              "width": 25,
            }
          }
        />
        <Text
          style={
            [
              {
                "color": "#3f7b82",
              },
              {
                "fontFamily": "Poppins-Bold",
                "fontSize": 14,
              },
              null,
            ]
          }
        >
          Item title 2
        </Text>
      </View>
    </BVLinearGradient>
  </View>
  <View
    style={
      [
        {
          "shadowColor": "#8b8490",
          "shadowOffset": {
            "height": 2,
            "width": 0,
          },
          "shadowOpacity": 0.25,
          "shadowRadius": 8,
        },
        {
          "backgroundColor": "#ffffff",
          "borderRadius": 8,
        },
      ]
    }
  >
    <BVLinearGradient
      angle={170}
      angleCenter={
        {
          "x": 0.75,
          "y": 0.75,
        }
      }
      colors={
        [
          4294967295,
          4294967295,
        ]
      }
      endPoint={
        {
          "x": 0.5,
          "y": 1,
        }
      }
      locations={null}
      startPoint={
        {
          "x": 0.5,
          "y": 0,
        }
      }
      style={
        {
          "alignItems": "center",
          "borderRadius": 8,
          "display": "flex",
          "flexDirection": "row",
          "gap": 10,
          "paddingHorizontal": 20,
          "paddingVertical": 10,
        }
      }
      useAngle={true}
    >
      <View
        accessibilityHint="Tap to mark as done"
        accessibilityLabel="Toggle Item title 3 done state"
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        role="button"
        style={
          {
            "alignItems": "center",
            "display": "flex",
            "flex": 1,
            "flexDirection": "row",
            "gap": 10,
          }
        }
      >
        <View
          accessibilityLabel="Item title 3 is not done"
          style={
            {
              "borderColor": "#3f7b82",
              "borderRadius": 50,
              "borderWidth": 2,
              "height": 25,
              "width": 25,
            }
          }
        />
        <Text
          style={
            [
              {
                "color": "#3f7b82",
              },
              {
                "fontFamily": "Poppins-Bold",
                "fontSize": 14,
              },
              null,
            ]
          }
        >
          Item title 3
        </Text>
      </View>
    </BVLinearGradient>
  </View>
</View>
`;
